core/src/services/ftp/writer.rs

```diff
@@ -37,30 +35,47 @@ pub struct FtpWriter {
 /// After we can use data stream, we should return it directly.
 impl FtpWriter {
     pub fn new(backend: FtpBackend, path: String) -> Self {
-        FtpWriter { backend, path }
+        FtpWriter {
+            backend,
+            path,
+            size: 0,
+        }
     }
 }
 
-/// # Safety
-///
-/// We will only take `&mut Self` reference for FtpWriter.
-unsafe impl Sync for FtpWriter {}
-
-impl oio::OneShotWrite for FtpWriter {
-    async fn write_once(&self, bs: Bytes) -> Result<()> {
+impl oio::Write for FtpWriter {
+    async unsafe fn write(&mut self, bs: oio::ReadableBuf) -> Result<usize> {
         let mut ftp_stream = self.backend.ftp_connect(Operation::Write).await?;
-        let mut data_stream = ftp_stream
-            .append_with_stream(&self.path)
-            .await
-            .map_err(parse_error)?;
-        data_stream.write_all(&bs).await.map_err(|err| {
+        let mut data_stream = match self.size {
```

Comment:
Please handle the `OpWrite::append()` correctly: if `append = false`, we should create a new file and keep writing on it, otherwise, we should append it directly.

However, I'm not sure if we can store `data_stream` inside `FtpWriter`. Would you like to have a try?

Reply:
> I'm not sure if I understand "create a new file and keep writing on it", is it similar to the implementation of `FsWriter`?

Yes. Keep the `data_stream` in `FtpWriter` so that we can write on the same connection instead of starting new connection for every write.

Reply:
> > I'm not sure if I understand "create a new file and keep writing on it", is it similar to the implementation of `FsWriter`?
> 
> Yes. Keep the `data_stream` in `FtpWriter` so that we can write on the same connection instead of starting new connection for every write.

It is somewhat difficult to store the `DataStream`, and there seems to be no direct way to access it.  I plan to pass the `ftp_stream` into the `FtpWriter` when constructing it, and the `data_stream` will be stored when the `FtpWriter` executes the `write` operation for the first time. Another reason for storing the `ftp_stream` is that it needs to be called `ftp_stream.finalize_put_stream(data_stream)` when closing.

Additionally, I can modify the implementation of `write` to first write to a temporary file, and then rename the file during `close` (only if `append = false`). This way, the `FtpWriter` can directly support the `append` mode. Do I need to do this?

Reply:
> Additionally, I can modify the implementation of `write` to first write to a temporary file, and then rename the file during `close` (only if `append = false`). This way, the `FtpWriter` can directly support the `append` mode. Do I need to do this?

Perfect! Please do this.

Reply:
> > Additionally, I can modify the implementation of `write` to first write to a temporary file, and then rename the file during `close` (only if `append = false`). This way, the `FtpWriter` can directly support the `append` mode. Do I need to do this?
> 
> Perfect! Please do this.

Bad news, it seems that storing `DataStream` is also not possible. Some crucial traits/structs liks `AsyncTlsStream` in `suppaftp` are private, so I have no way to reference these types. I saw this point mentioned in the issue of the `suppaftp` repo as well: https://github.com/veeso/suppaftp/issues/60, but it seems that there hasn't been any progress for a long time.